### ISSUE TITLE
Ability to get the worker processing a job

### ISSUE THREAD

**Issue by btmdave** [OptimalBits/bull#173]
We're in the process of moving from kue to bull, because quite frankly it works great!  One feature that was not implemented in kue, but would be very useful for larger deployments, is to know which server is processing a particular job.

For example, if we have an encoding farm with 10+ servers each processing encoding jobs, it would be useful to know which server is processing what.   In Kue, we would update the data for a job with the hostname.  This wasn't ideal though, because if we wanted to aggregate by hostname, you basically have to parse all jobs, which really doesn't scale.   

I think it would be useful to have an additional set that's created based on the server processing jobs, for example there could be set that has `bull:queue-name:hosts:hostname` which includes the jobIds for each queue being processed by that host.    

Maybe this is a little out of scope, but I think it's something worth thinking about - having more detailed reports into what's going on across a cluster of workers.

**Comment by manast** [comment]
I like that idea. And in fact would be really useful with some kind of UI. I have been thinking quite seriously to create an UI as well to get a complete overview of all the jobs, queues and workers.

**Comment by albertjan** [comment]
:+1: We've run into that as well I was thinking it might be possible to make the lock we take when we're processing a job include the server name or something. That way we could see which server has the job. But we also might want to think of a adding the name to the job hash after we're done so we can also see which server processed the job after the fact.

**Comment by btmdave** [comment]
That would be awesome.  If you want help doing the UI, maybe as a separate repo? Let me know.  I would imagine that it would be good for the UI, to be able to utilize built in methods within bull, as oppose to working with redis itself.  Off the bat, here's what I would I find useful in a general interface.

- **Dashboard**
*ability to group by All or queue type, i.e., Audio Encoding, Video Encoding and Server, i.e.,  worker001, worker002, etc.*

 - Waiting (count) - *view*
 - Active (count) - *view*
 - Delayed (count) - *view*
 - Completed (count) - *view*
 - Failed (count) - *view*

- **View**
 *ability to group by same as above*
 - List of jobs
    - jobId
    - title (if defined)
    - status
    - created date
    - progress (if defined)
    - details (expands shows all job data)
    - actions (pause, resume, cancel)

Basically this would give us the ability to get a decent snapshot of what's going on by queue type and by server, then some more job specific data.   Maybe later, some higher level remote control abilities, like pausing and resume a specific queue.

**Comment by manast** [comment]
great feedback. I could start with the bare bones of the project and then invite everybody else to contribute, in that way we could have something working faster. By the way, my personal preference for the UI framework would be angular, since thats what I am more proficient with... hope thats ok.

**Comment by btmdave** [comment]
Nice, yeah I was going to recommend angular also :)  Sounds great

**Comment by albertjan** [comment]
You can't pause/resume a job. Those actions should probably go on the queue view. 

There should also probably be a retry button that moves a job from failed/completed to the wait list. 

I'll pitch in where possible. @remkoboschker is probably also interested.

**Comment by kulbirsaini** [comment]
This is exciting stuff. Let me know if I can be of any help at all. I am not good at UI part though.

@manast Also, can we have a gitter/irc channel setup?

**Comment by evanhuang8** [comment]
Any updates on this? I have quite some free time over the weekend and I would love to contribute

**Comment by manast** [comment]
I have started this project, but ran out of time, I have some stuff to push to this repo: https://github.com/OptimalBits/bull-ui
I need a few hours to make something worth pushing though, is mostly file structure and so.

**Comment by manast** [comment]
I made a push with a basic structure. You are welcome to contribute, the next part is to implement the APIs and of course make a nice design :)

**Comment by evanhuang8** [comment]
@manast looks great! I actually jumped the gun yesterday and started making my own version haha. Nothing in the front end, but I wrote a bunch of basic API end points: 

https://github.com/evanhuang8/Toureiro

Checkout the stuff in `lib/models/`, let me know what you think. Maybe we can merge the two efforts together?

**Comment by manast** [comment]
@evanhuang8 I think that your library can be used to implement the back end methods in bull-ui, so if you make it a public npm module I can depend on it.

**Comment by evanhuang8** [comment]
@manast I can just copy them and make a pull request to bull-hi, no point to keep them separate :)

**Comment by evanhuang8** [comment]
@manast So I ended up writing the whole thing since I needed it, now with a simple UI using React. Check it out at: https://www.npmjs.com/package/toureiro.

**Comment by albertjan** [comment]
Haha thats awesome I will check it out tomorrow.

**Comment by manast** [comment]
great stuff. The UI will need some love but it looks promising, time for me to learn react I guess :)

**Comment by evanhuang8** [comment]
Thought people would appreciate a little throw back style haha...
